core/app/models/spree/line_item.rb

```diff
@@ -138,49 +114,68 @@ def options=(options={})
       opts = options.dup # we will be deleting from the hash, so leave the caller's copy intact
 
       currency = opts.delete(:currency) || order.try(:currency)
+      set_price(currency, opts)
+
+      self.assign_attributes opts
+    end
 
+    private
+    # Sets the quantity to zero if it is nil or less than zero.
+    def invalid_quantity_check
+      self.quantity = 0 if quantity.nil? || quantity < 0
+    end
+
+    # Sets this line item's  tax category, price, cost price, and currency from
+    # its variant if they are nil and a variant is present.
+    def copy_variant_attributes
+      return unless variant
+      self.tax_category ||= variant.tax_category
+      self.cost_price ||= variant.cost_price
+      self.currency ||= variant.currency
+      set_price
+    end
+
+    # Copy the price including any price modifiers from the variant
+    def set_price(currency = nil, options = {})
```

Comment:
I would be more explicit in the naming here. "Set the price from what?"
Why not name this method `set_price_from_variant`?

It's a little bit weird, that this method does two things, from two different callers. Ehy not let this little fellow just set the price? And the currency can be set from somewhere else? Inside `def options= ` for instance? Then you can spare the first argument and it's more clear what this method does.